src/audio/module_adapter/module_adapter_ipc4.c

```diff
@@ -38,19 +38,39 @@ int module_adapter_init_data(struct comp_dev *dev,
 			     const struct comp_ipc_config *config,
 			     const void *spec)
 {
-	if (dev->drv->type == SOF_COMP_MODULE_ADAPTER) {
-		const struct ipc_config_process *ipc_module_adapter = spec;
+	const struct ipc_config_process *cp = spec;
+	const struct ipc4_base_module_extended_cfg *cfg = (void *)cp->data;
+	size_t cfgsz = cp->size;
 
-		dst->init_data = ipc_module_adapter->data;
-		dst->size = ipc_module_adapter->size;
-		dst->avail = true;
+	assert(dev->drv->type == SOF_COMP_MODULE_ADAPTER);
```

Comment:
It would seem this is ok. Most modules use "DECLARE_MODULE_ADAPTER()" to register the driver (which sets the type correctly) and library_manager calls "declare_dynamic_module_adapter(drv, SOF_COMP_MODULE_ADAPTER, ..)". So I don't see any use for IPC4 where the type would not be set to adapter. It still bugs me a bit a specific branch was added to cover this case and I can't find from git history any use for this, but maybe this was some variant of loadable modules. @jxstelter @ranj063 @RanderWang  @softwarecki @pjdobrowolski please do comment if I got this wrong.

Reply:
That was my read, yeah.  The original had a check for that that then set the init_data unconditionally, which seems a little dangerous (what if it's not?  what's in the data and how do we prevent treating it like pin config?".  I grepped around and couldn't find any situations where it would happen, so I replaced it with an assert to be sure, figuring that CI would tell me if I missed something.

Reply:
Is this really fatal ? I think its probably survivable where we return the error up the stack for this module/pipeline only and allow other pipelines/streams to continue.